Chris process proposal

cwjnkins opened the pull request:
This PR fleshes out the model of the implementation in `LibraBFT.Impl`, specifically `processProposalMsgM` and its dependencies.

cwjnkins commented:
It also removes `LibraBFT.Impl.OBM.Prelude` -- I thought the functionality belonged in `RWST.agda` instead.

mark-moir commented on LibraBFT/ImplShared/Consensus/Types.agda:
```diff
@@ -31,11 +31,13 @@ open import Data.String using (String)
+  open import LibraBFT.ImplFake.NetworkMsg                       public
+  open import LibraBFT.ImplShared.Base.Types                     public
+  open import LibraBFT.ImplShared.Consensus.Types.EpochIndep     public
+  open import LibraBFT.ImplShared.Consensus.Types.MetaEpochIndep public
```
This file doesn't exist, and doesn't appear to be needed.

cwjnkins replied:
It is the new home for the definition of `Output`. I moved `MetaEpochIndep` to the `ImplShared` directory, but forgot to track it -- this has been fixed now.

mark-moir commented on LibraBFT/ImplShared/Consensus/Types/EpochIndep.agda:
```diff
@@ -560,19 +591,6 @@ module LibraBFT.ImplShared.Consensus.Types.EpochIndep where
     UnexpectedRound       : Round → Round →      VoteReceptionResult
     VRR_TODO              :                      VoteReceptionResult
 
-  data Output : Set where
```
I guess you intended to move this somewhere else, which makes sense, but it doesn't seem to be anywhere now.

cwjnkins replied:
It has been added to `LibraBFT.ImplShared/Consensus/Types/MetaEpochIndep.agda`

mark-moir commented on Optics/Functorial.agda:
```diff
@@ -50,6 +51,10 @@ module Optics.Functorial where
 
+  set? : ∀{S A} → Lens S (Maybe A) → A → S → S
+  set? l a s = s [ l := just a ]
+  syntax set? l a s = s [ l ?= a ]
```
Any reason not to call this operator `?~` instead of `?=`?  I think that's the Haskell spelling.  I'm also wondering whether we could/should do:
```
  syntax set? l a s = s & l ?~ a 
```
for better consistency with Haskell syntax.  (Similarly for `set` above.)  These aren't a huge detail, but the closer the model is to the implementation, the easier is too review, and the less risk there is of overlooking a difference.

cwjnkins replied:
In Haskell, the `&` operator is function application. We I propose we rename `_|>_` from the Agda standard library's `Function` module (which does the same thing) to `_&_`, and define syntax for `l ?= a : S -> S`, to make the correspondence as close as possible.

haroldcarr replied:
This might be useful:

[https://hackage.haskell.org/package/lens-5.0.1/docs/Control-Lens-Operators.html](https://hackage.haskell.org/package/lens-5.0.1/docs/Control-Lens-Operators.html)

mark-moir replied:
@cwjnkins that sounds reasonable to me, but I'm not going to hold up this PR for it.  It would be great if you could experiment with this and, if it makes sense, sneak it into another PR soon.

mark-moir replied:
Oh, never mind, I see you've already done it.  Great!

mark-moir commented:
Overall looks very good @cwjnkins.  I do have a few comments and suggestions.  It's especially important to address the ones that break things so that `./Scripts/run-everything y` doesn't succeed.  Thanks!

cwjnkins commented on LibraBFT/ImplShared/Consensus/Types/EpochIndep.agda:
```diff
@@ -348,6 +371,12 @@ module LibraBFT.ImplShared.Consensus.Types.EpochIndep where
              (pmProposal ∷ pmSyncInfo ∷ [])
   postulate instance enc-ProposalMsg : Encoder ProposalMsg
 
+  -- The implementation exits with an error if a proposal message contains no
```
Actually, @haroldcarr changed the Haskell implementation to no longer behave this way (exiting with an error is how the Rust implementation behaves), so this comment can be removed.

haroldcarr replied:
The Haskell now returns `(Maybe Author)` too.
That comment should be deleted.

cwjnkins commented:
@mark-moir I've just pushed a commit addressing your review. Do you think that `MetaEpochIndep` belongs in `ImplShared`, or should I move that completely to `Impl` and restore the older version of `Output` for `FakeImpl`?

haroldcarr commented on LibraBFT/ImplShared/Consensus/Types.agda:
```diff
@@ -191,10 +197,12 @@ module LibraBFT.ImplShared.Consensus.Types where
   rmGetBlockStore : (rm : RoundManager) → BlockStore (α-EC-RM rm)
   rmGetBlockStore rm = (₋rmWithEC rm) ^∙ (epBlockStore (α-EC-RM rm))
 
+  rmSetBlockStore : (rm : RoundManager) → BlockStore (α-EC-RM rm) → RoundManager
```
This does not appear to be used.
Let's delete it if not needed.

cwjnkins replied:
I make use of this in a branch that extends this one, and it will be needed to implement `BlockStore.executeAndInsertBlockM` (since we cannot write a lens for the block store). To leave it in that branch though would might lead to further progress on developing the implementation model adding it independently, leading to an unnecessary (though minor) merge conflict.

haroldcarr replied:
If you plan on using it then no need to remove.

haroldcarr commented on LibraBFT/ImplShared/Consensus/Types/EpochIndep.agda:
```diff
@@ -307,6 +310,26 @@ module LibraBFT.ImplShared.Consensus.Types.EpochIndep where
+      ₋vpBlock : Block
+      -- ₋vpNextEpochState : Maybe EpochState
+  open VoteProposal public
+  unquoteDecl vpBlock = mkLens (quote VoteProposal) (vpBlock ∷ [])
```
please align `vpBlock` on two lines

haroldcarr commented on LibraBFT/ImplShared/Consensus/Types/EpochIndep.agda:
```diff
@@ -307,6 +310,26 @@ module LibraBFT.ImplShared.Consensus.Types.EpochIndep where
+      ₋msvpSignature : Maybe Signature
+  open MaybeSignedVoteProposal public
+  unquoteDecl msvpVoteProposal msvpSignature = mkLens (quote MaybeSignedVoteProposal)
+              (msvpVoteProposal ∷ msvpSignature ∷ [])
```
please align the lens names

haroldcarr commented on LibraBFT/ImplShared/Consensus/Types/MetaEpochIndep.agda:
```diff
@@ -0,0 +1,58 @@
+   Copyright (c) 2021, Oracle and/or its affiliates.
+   Licensed under the Universal Permissive License v 1.0 as shown at https://opensource.oracle.com/licenses/upl
+-}
+open import Optics.All
```
alphabetic imports

haroldcarr commented:
Overall - looks great.

cwjnkins commented on LibraBFT/Impl/Consensus/BlockStorage/BlockStore.agda:
```diff
@@ -24,6 +21,8 @@ module LibraBFT.Impl.Consensus.BlockStorage.BlockStore where
 open RWST-do
 
 postulate
+  executeAndInsertBlockM : Block → LBFT (Unit ⊎ ExecutedBlock)
```
My mistake, I must have wrote this before I started using the fake `ErrLog` (defined to be `Unit` for now, but to be changed soon following recent discussions.)

cwjnkins commented on LibraBFT/Impl/Consensus/RoundManager.agda:
```diff
@@ -95,6 +98,106 @@ addVoteM now vote = do
+      else ok true
+
+  processProposalM proposal = do
+    _rm ← get
```
This was an attempt to call attention to the fact that `_rm` is not bound in the Haskell implementation. I will add a comment here explaining the issue, but is there a naming convention we could use for these situations (which arise because we do not have dependent lenses, so some fields of `RoundManager` must be set this way)?

haroldcarr replied:
No convention (yet?).
I would just do the `s <- get` or `gets` would be even better.

cwjnkins commented on LibraBFT/Impl/Consensus/SafetyRules/SafetyRules.agda:
```diff
@@ -0,0 +1,121 @@
+
+constructAndSignVoteM-continue1 voteProposal validatorSigner proposedBlock safetyData0 =
+  verifyQcM (proposedBlock ^∙ bQuorumCert) ∙?∙ λ _ → do
+    validatorVerifier ← gets rmGetValidatorVerifier
```
In our `RoundManager`, the `ValidatorVerifier` is mentioned in the proof field `₋rmEC-correct`, so we do not have a lens for this field.

haroldcarr replied:
Oh yeah.   I like the use of `gets` in this situation.

cwjnkins commented on LibraBFT/Impl/Consensus/SafetyRules/SafetyRules.agda:
```diff
@@ -0,0 +1,121 @@
+      constructLedgerInfoM proposedBlock (Crypto.hashVD voteData) ∙?∙ λ ledgerInfo → do
+        let signature = ValidatorSigner.sign validatorSigner ledgerInfo
+            vote      = Vote.newWithSignature voteData author ledgerInfo signature
+        lSafetyData ∙= (safetyData1 [ sdLastVote ?= vote ])
```
Per a previous discussion with @mark-moir, we do not have the equivalent of an `obmMakeClassy` such that `pssSafetyData` can work in this context (its type is `Lens PersistentSafetyStorage SafetyData`). The solution we were discussing (but IIRC has not been fully resolved) is to use names such as `pssSafetyData-RoundManager`.

haroldcarr replied:
Right.
I hate that suggested name.
How about `rm_pssSafetyData` ?

cwjnkins commented on LibraBFT/Impl/Types/ValidatorSigner.agda:
```diff
@@ -15,5 +15,5 @@ open import Optics.All
 
-sign : ∀ {A : Set} → ValidatorSigner → A → Signature
-sign = {!!}
+sign : {C : Set} ⦃ ws : WithSig C ⦄ → ValidatorSigner → C → Signature
```
I am using the same variable name as the parameter to the `WithSig`record (I think this was a function we both independently added).

cwjnkins commented on LibraBFT/Impl/Consensus/SafetyRules/SafetyRules.agda:
```diff
@@ -0,0 +1,121 @@
+    then bail unit -- error: incorrect preferred round, QC round does not match preferred round
+    else do
+      updated ← grd‖ twoChainRound >? preferredRound
+                     ≔ pure (safetyData [ sdPreferredRound := twoChainRound ])
```
Yes, I have done this.

cwjnkins commented on LibraBFT/Impl/Consensus/RoundManager.agda:
```diff
@@ -95,6 +98,106 @@ addVoteM now vote = do
+  processProposalMsgM : Instant → {- Author → -} ProposalMsg → LBFT Unit
+  processProposalMsgM now {- from -} pm
+     with pm ^∙ pmProposer
+  ...| nothing = pure unit -- errorExit "ProposalMsg does not have an author"
```
That is a good question. For `processProposalMsgM`, I have been using the output (i.e., whether it has any or not) as an indicator of whether it succeeded for failed. If that is true for the other top-level functions of type `LBFT Unit`, then perhaps this suffices?

haroldcarr replied:
In Haskell:

```
data Output a
  = BroadcastEpochChangeProof LogControl EpochChangeProof     NodesInOrder
  | BroadcastProposal         LogControl (ProposalMsg a)      NodesInOrder
  | BroadcastSyncInfo         LogControl SyncInfo             NodesInOrder
  | BroadcastVote             LogControl VoteMsg              NodesInOrder
  | LogErr                               (ErrLog a)
  | LogInfo                   LogControl (InfoLog a)
  | ReconfigLocalEpochChange  LogControl ReconfigEventEpochChange
  | SendBRP                   LogControl Author               (BlockRetrievalResponse a)
  | SendVote                  LogControl VoteMsg              NodesInOrder
  | SetTimeout                LogControl Duration             Round
  deriving (Eq, Show)
```
so the errors DO show up in the output.  All the "top level" functions in `RoundManager` return `LBFT Unit` and are expected to catch and report errors from below.

So I do think it suffices.

Do we want real error types rather than the existing fake unit?

haroldcarr replied:
What this means, is that all the places like:

```
processProposalMsgM : Instant → {- Author → -} ProposalMsg → LBFT Unit
processProposalMsgM now {- from -} pm =
  case pm ^∙ pmProposer of λ where
    nothing        → pure unit -- log: info: proposal with no author
    (just pAuthor) →
      ensureRoundAndSyncUpM now (pm ^∙ pmProposal ∙ bRound) (pm ^∙ pmSyncInfo)
                            pAuthor true >>= λ where
      (inj₁ _)     → pure unit -- log: error: <propagate error>
      (inj₂ true)  → processProposalM (pm ^∙ pmProposal)
      (inj₂ false) → do
        currentRound ← use (lRoundState ∙ rsCurrentRound)
        pure unit -- log: info: dropping proposal for old round
```
where we are returning `pure unit` should do like Haskell and log a specific error into `Output`

cwjnkins replied:
This is a discussion I would want @mark-moir involved in. My guess is that we would definitely want this eventually, but I do not its absence being being an immediate impediment to the proof effort.

mark-moir replied:
Thanks guys, this is an important discussion.  If we construct proofs with the assumption (accurate for our partial model for now) that, for example, `processProposalMsg` produces either no `Output`s or exactly one `Output` which is a `SendVote`, these proofs would break when we introduce more `Output` types later.

On the other hand, we don't want to spend time now accurately modeling logged messages and errors, as we're more interested in correctness in the short term, rather than proving that certain circumstances result in certain error messages, etc.

So, my feeling is that we should make the model accurately reflect the implementation in terms of what *kinds* of outputs it sends, but *not* bother for now with details like actual log/error messages.

As mentioned before, `unit` shows up in our initial proofs for various things, and I'd much see something like `fakeError` in proofs than `unit`.  So maybe we could define:
```
FakeError : Set
postulate
  fakeError : FakeError
```
and have an `Output` constructor like `LogErr FakeError`?

That way the proofs would adapt better when we do refine the model to log actual errors, and we wouldn't fall into the trap mentioned above.

I think we should not hold up this PR for this issue, but we should resolve the discussion and implement what we decide soon, before spending time developing more proofs that will break when we refine the model.

mark-moir replied:
FYI, I have done something along these lines in a pull request I have just sent (#47).

haroldcarr commented:
The updates look great.

The main thing we need to do is log ERRORS (not INFO).
But I think that should be done in a separate pull request.

I approve for merging.

cwjnkins commented:
Noticed that `NetworkMsg.agda` still lived in `LibraBFT.ImplFake`, moved it to `LibraBFT.ImplShared`.

mark-moir commented:
> @mark-moir I've just pushed a commit addressing your review. Do you think that `MetaEpochIndep` belongs in `ImplShared`, or should I move that completely to `Impl` and restore the older version of `Output` for `FakeImpl`? The benefit of the former case is that `FakeImpl` could be used for proofs connecting meta-data to the behavior of the fake implementation, before attempting to do so for the real model of the implementation.

I agree with the suggested approach (especially because it's what you've already done).  We can always change this later if it becomes an issue, but in the meantime, less duplication is good.

mark-moir commented on LibraBFT/Impl/OBM/Prelude.agda:
```diff
@@ -1,31 +0,0 @@
-{- Byzantine Fault Tolerant Consensus Verification in Agda, version 0.9.
```
@cwjnkins, you resolved the conversation, but didn't record why.  I think it's because you replaced these with Agda standard library functions, but I tend to agree that we should keep this faithful to the implementation (e.g., suppose we got the definitions wrong in the implementation, but using the correct version from the Agda library masked the issue).

mark-moir commented on LibraBFT/ImplShared/Consensus/Types.agda:
```diff
@@ -153,26 +155,16 @@ module LibraBFT.ImplShared.Consensus.Types where
      -- RoundManager directly
   open RoundManager public
 
+  -- TODO-2: We would need dependent lenses to have a lens from RoundManager to
```
One thing to note is that we have quite a few `GetterNoFunctor` definitions in the Haskell code.  Because these are `Getter`-only, we _can_ have the equivalent in our context.  I think this would greatly reduce the number of cases in which the Agda model looks different to the Haskell code.  Something to think about; I'm not holding up this PR for this.

mark-moir commented on LibraBFT/ImplShared/Util/RWST.agda:
```diff
@@ -135,13 +143,37 @@ module LibraBFT.ImplShared.Util.RWST (ℓ-State : Level) where
   -- Derived Functionality
-
-  maybeMP : RWST Ev Wr St (Maybe A) → B → (A → RWST Ev Wr St B)
+  maybeSM : RWST Ev Wr St (Maybe A) → RWST Ev Wr St B → (A → RWST Ev Wr St B) → RWST Ev Wr St B
```
Oh, I see I was wrong about replacing these with standard library functions.  I agree with @haroldcarr that it's better to keep these definitions in the same place as the Haskell code.

mark-moir commented:
@cwjnkins and @haroldcarr : great stuff, thanks!  We do have a few unresolved conversations, but I'm merging now and we should go through those and decide what to do about them and when.
